lib/utils.js

```diff
@@ -1,27 +1,74 @@
+const path = require('path');
 const YAML = require('js-yaml');
+const { yamlAST, loc } = require('@fmvilas/pseudo-yaml-ast');
+const jsonAST = require('json-to-ast');
+const jsonParseBetterErrors = require('../lib/json-parse');
+const ParserError = require('./errors/parser-error');
 
-module.exports.toJS = (asyncapiYAMLorJSON) => {
+const utils = module.exports;
+
+utils.toJS = (asyncapiYAMLorJSON) => {
   if (!asyncapiYAMLorJSON) {
-    throw new Error(`Document can't be null, false or empty.`);
+    throw new ParserError({
+      type: 'null-or-falsey-document',
+      title: `Document can't be null or falsey.`,
+    });
   }
 
-  if (typeof asyncapiYAMLorJSON === 'object') {
-    return asyncapiYAMLorJSON;
+  if (asyncapiYAMLorJSON.constructor && asyncapiYAMLorJSON.constructor.name === 'Object') {
+    return {
+      initialFormat: 'js',
+      parsedJSON: asyncapiYAMLorJSON,
+    };
+  }
+  
+  if (typeof asyncapiYAMLorJSON !== 'string') {
+    throw new ParserError({
+      type: 'invalid-document-type',
+      title: 'The AsyncAPI document has to be either a string or a JS object.',
+    });
   }
 
-  try {
-    return JSON.parse(asyncapiYAMLorJSON);
-  } catch (e) {
+  if (asyncapiYAMLorJSON.trimLeft().startsWith('{')) {
```

Comment:
This! 🤦 It is so easy to understand if a document is JSON or not. Don't know why I was trying a nested try/catch. Also, the nested try/catch had a problem: when it fails parsing json **and** yaml, you don't know if it's because it's invalid json or it's invalid yaml. It's impossible to guess the intent. The way it was, it was favoring YAML over JSON.